src/ripple/core/sql_backend/SQLInterface_sqlite.cpp

```diff
@@ -0,0 +1,3747 @@
+
+    virtual boost::variant<Transaction::pointer, bool>
+    loadTransaction(
+        SQLDatabase_* db,
+        Application& app,
+        uint256 const& id,
+        boost::optional<ClosedInterval<uint32_t>> const& range,
+        error_code_i& ec) override;
+
+    virtual bool
+    checkDBSpace(SQLDatabase& txDb, Config const& config, beast::Journal& j)
+        override;
+
+    virtual std::pair<PublicKey, SecretKey>
+    loadNodeIdentity(SQLDatabase& db) override;
+
+    virtual void
+    databaseBodyDoPut(
+        SQLDatabase& conn,
+        std::string& data,
+        std::string& path,
+        std::uint64_t fileSize,
+        std::uint64_t& part,
+        std::uint16_t const maxRowSizePad) override;
+
+    virtual void
+    databaseBodyFinish(SQLDatabase& conn, std::ofstream& fout) override;
+
+    virtual void
+    addValidatorManifest(SQLDatabase& db, std::string const& serialized)
+        override;
+
+    virtual void
+    loadPeerReservationTable(
+        SQLDatabase& conn,
+        beast::Journal& j,
+        std::unordered_set<PeerReservation, beast::uhash<>, KeyEqual>& table)
+        override;
+
+    virtual void
+    insertPeerReservation(
+        SQLDatabase& conn,
+        PublicKey const& nodeId,
+        std::string const& description) override;
+
+    virtual void
+    deletePeerReservation(SQLDatabase& conn, PublicKey const& nodeId) override;
+
+    virtual void
+    readArchiveDB(
+        SQLDatabase& db,
+        std::function<void(std::string const&, int)> const& func) override;
+
+    virtual void
+    insertArchiveDB(
+        SQLDatabase& db,
+        LedgerIndex shardIndex,
+        std::string const& url) override;
+
+    virtual void
+    deleteFromArchiveDB(SQLDatabase& db, LedgerIndex shardIndex) override;
+
+    virtual void
+    dropArchiveDB(SQLDatabase& db) override;
+
+    virtual int
+    getKBUsedAll(SQLDatabase& db) override;
+
+    virtual int
+    getKBUsedDB(SQLDatabase& db) override;
+
+    virtual void
+    readPeerFinderDB(
+        SQLDatabase& db,
+        std::function<void(std::string const&, int)> const& func) override;
+
+    virtual void
+    savePeerFinderDB(
+        SQLDatabase& db,
+        std::vector<PeerFinder::Store::Entry> const& v) override;
+
+private:
+    bool
+    loadLedgerInfo(
+        SQLDatabase_* db,
+        SQLInterface::SQLLedgerInfo& info,
+        beast::Journal const& j,
+        std::string const& sqlSuffix);
+
+    std::string
+    transactionsSQL(
+        Application& app,
+        beast::Journal& j,
+        AccountID const& account,
+        std::string selection,
+        std::int32_t minLedger,
+        std::int32_t maxLedger,
+        bool descending,
+        std::uint32_t offset,
+        int limit,
+        int limit_used,
+        bool binary,
+        bool count,
+        bool bUnlimited);
+};
+
+static bool
+exists(SQLDatabase& db)
+{
+    if (db)
+    {
+        SQLDatabase_sqlite* sdb = dynamic_cast<SQLDatabase_sqlite*>(&*db);
```

Comment:
The `SQLDatabase` type is odd. It's a pointer in disguise. Working with a type thats secretly a pointer is error prone and hard to read. I think we can get rid of the `SQLDatabase_` and just use `std::unique_ptr<SQLDatabase>` when we need a pointer to a `SQLDatabase`. `&*db` is pretty confusing.